### PR TITLE
fix soft affinities/constraints

### LINKED ISSUE #1374: soft affinity and constraints not working properly together (1.0.0-rc3)
if we launch a docker container with the following soft affinity/constraints, swam throws the following error:  "Error response from daemon: unable to find a node that satisfies mitre.pool==B"

```
-e constraint:mitre.environment==DEV
-e constraint:mitre.pool==B
-e affinity:container==~busybox
```
This occurs when busybox is already running on pool A, and we are trying to start busybox on pool B as well. What should happen here is that even though the soft affinity is being met by r10a-dockersb1, the container should run on r10a-dockersb2 in our environment because the constraints should outweigh the soft affinity.  Here are the details about our env:

```
[bamboo@r10a-dockersb1 ~]$ docker info
Containers: 6
Images: 9
Role: primary
Strategy: spread
Filters: health, port, dependency, affinity, constraint
Nodes: 2
 r10a-dockersb1.mitre.org: r10a-dockersb1.mitre.org:2376
  â Containers: 4
  â Reserved CPUs: 0 / 2
  â Reserved Memory: 0 B / 3.888 GiB
  â Labels: executiondriver=native-0.2, kernelversion=3.10.0-229.14.1.el7.x86_64, mitre.environment=SND, mitre.f5=B0k103-dltm1.mitre.org, mitre.pool=A, operatingsystem=Red Hat Enterprise Linux Server 7.1 (Maipo), storagedriver=devicemapper
 r10a-dockersb2.mitre.org: r10a-dockersb2.mitre.org:2376
  â Containers: 2
  â Reserved CPUs: 0 / 2
  â Reserved Memory: 0 B / 3.888 GiB
  â Labels: executiondriver=native-0.2, kernelversion=3.10.0-229.14.1.el7.x86_64, mitre.environment=SND, mitre.f5=B0k103-dltm1.mitre.org, mitre.pool=B, operatingsystem=Red Hat Enterprise Linux Server 7.1 (Maipo), storagedriver=devicemapper
CPUs: 6
Total Memory: 11.66 GiB
Name: r10a-dockersb2.mitre.org
```
Here are the versions:
```
[bamboo@r10a-dockersb1 ~]$ docker version
Client:
 Version:      1.9.0-rc4
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   e6f5a3c
 Built:        Sat Oct 31 05:43:43 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      swarm/1.0.0
 API version:  1.21
 Go version:   go1.5.1
 Git commit:   d395fa5
 Built:
 OS/Arch:      linux/amd64

[bamboo@r10a-dockersb1 ~]$ docker-compose --version
docker-compose version: 1.5.0rc3
```

It seems to us that the soft affinity is being treated more like a "hard" affinity because if we remove the affinity from the equation then it works fine, for example...
```
-e constraint:mitre.environment==DEV
-e constraint:mitre.pool==B
```

### PR DESCRIPTION
**Pull request by vieux**
Fix #1374

### CODE REVIEW

**Comment by vieux** [comment]
tests are ok locally, something is not right with the CI

**Comment by vieux** [comment]
ping @docker/swarm-maintainers

File: scheduler/filter/filter.go [hunk 1]
```diff
@@ -55,15 +55,28 @@ func New(names []string) ([]Filter, error) {
 
 // ApplyFilters applies a set of filters in batch.
 func ApplyFilters(filters []Filter, config *cluster.ContainerConfig, nodes []*node.Node) ([]*node.Node, error) {
-	var err error
+	candidates, err := applyFilters(filters, config, nodes, true)
```
Comment (aluzzardi) [inline scheduler/filter/filter.go]: shouldn't we do hard first, soft later?

**Comment by aluzzardi** [comment]
I'm not sure I understand the fix.

Couldn't we just skip all soft at the beginning, then take the resulting list of candidates and apply all softs?

File: scheduler/filter/filter.go (hunk 1 above)
Comment (abronan) [inline scheduler/filter/filter.go]: I'm wondering the same thing.

Comment (vieux) [inline scheduler/filter/filter.go]: that's not at all what we did before.
currently we try with hard and soft and if it doesn't work, we drop to soft (and try only with the hard)

**Comment by jimmyxian** [comment]
From my understanding, the sequence of filter under this PR is:
```
HealthFilter
PortFilter
DependencyFilter
AffinityFilter
SoftAffinityFilter
ConstraintFilter
SoftConstraintFilter
```
In my opinion, we should do hard filter first. If we get more than one nodes after hard filters, then try do soft filters.
```
HealthFilter
PortFilter
DependencyFilter
AffinityFilter
ConstraintFilter
SoftAffinityFilter
SoftConstraintFilter
```

**Comment by dongluochen** [comment]
I agree with the hard first approach. I think we can also provide a `preview` run with debug information. The command doesn't take any real effect but return the candidate set, or how many candidates are available. Ops people can understand what filters may work, or what is going to happen.

**Comment by vieux** [comment]
If we change to hard first, it's going of change the behaviour of swarm.
See this sentence https://github.com/docker/swarm/tree/master/scheduler/filter#soft-affinitiesconstraints

This PR doesn't change anything, just fixes an issue (#1374)

**Comment by vieux** [comment]
@jimmyxian for this PR I would say

```
HealthFilter
PortFilter
DependencyFilter
AffinityFilter(with soft included)
ConstraintFilter(with soft included)
```

If this returns no node, we retry with:

```
HealthFilter
PortFilter
DependencyFilter
AffinityFilter
ConstraintFilter
```

**Comment by jimmyxian** [comment]
@vieux Got it.
LGTM to fix https://github.com/docker/swarm/issues/1374

**Comment by abronan** [comment]
@vieux `docker run - reschedule with soft-image-affinity` is failing, I guess we should update the test.

**Comment by vieux** [comment]
@abronan I just rebased, LGTM

**Comment by dongluochen** [comment]
LGTM.

**Pull request merged**